R/seasonal_burden_levels.R

```diff
@@ -106,47 +109,64 @@ seasonal_burden_levels <- function(
-  season_observations_and_weights <- weighted_seasonal_tsd |>
-    dplyr::select(-c("year", "time")) |>
+  peak_seasonal_tsd <- seasonal_tsd |>
     dplyr::filter(.data$observation >= disease_threshold) |>
     dplyr::slice_max(.data$observation, n = n_peak, with_ties = FALSE, by = "season")
 
-  # Run quantiles_fit function
-  quantiles_fit <- season_observations_and_weights |>
-    dplyr::select("observation", "weight") |>
-    fit_quantiles(weighted_observations = _, conf_levels = conf_levels, ...)
+  # main level function
+  main_level_fun <- function(seasonal_tsd) {
+    # Add weights and remove current season to get predictions for this season
+    weighted_seasonal_tsd <- seasonal_tsd |>
+      dplyr::filter(.data$season != max(.data$season)) |>
+      dplyr::mutate(year = purrr::map_chr(.data$season, ~ stringr::str_extract(.x, "[0-9]+")) |>
+                      as.numeric()) |>
+      dplyr::mutate(weight = decay_factor^(max(.data$year) - .data$year)) |>
+      dplyr::select(-c("year", "time"))
+
+    # Run quantiles_fit function
+    quantiles_fit <- weighted_seasonal_tsd |>
+      dplyr::select("observation", "weight") |>
+      fit_quantiles(weighted_observations = _, conf_levels = conf_levels, ...)
+
+    # If method intensity_levels was chosen; use the high level from the `fit_quantiles` function as the high
+    # level and the disease_threshold as the very low level. The low and medium levels are defined as the relative
+    # increase between the very low level and high level.
+    results <- switch(method,
+      peak_levels = {
+        model_output <- append(quantiles_fit, list(season = max(seasonal_tsd$season)), after = 0)
+        model_output$values <- stats::setNames(c(disease_threshold, model_output$values),
+                                               c("very low", "low", "medium", "high"))
+        model_output <- append(model_output, list(disease_threshold = disease_threshold))
+      },
+      intensity_levels = {
+        level_step_log <- pracma::logseq(disease_threshold, quantiles_fit$values, n = 4)
+        model_output <- list(
+          season = max(seasonal_tsd$season),
+          high_conf_level = quantiles_fit$conf_levels,
+          values = stats::setNames(level_step_log, c("very low", "low", "medium", "high")),
+          par = quantiles_fit$par,
+          obj_value = quantiles_fit$conf_levels,
+          converged = quantiles_fit$converged,
+          family = quantiles_fit$family,
+          disease_threshold = disease_threshold
+        )
+      }
+    )
+    return(results)
+  }
+  # Select seasons for output based on only_current_season input argument
+  if (only_current_season == FALSE) {
+    # Group seasons to get results for all seasons available
+    unique_seasons <- unique(rev(peak_seasonal_tsd$season))
+    season_groups <- purrr::map(seq_along(unique_seasons), ~ unique_seasons[1:.x]) |>
+      purrr::accumulate(union)
```

Comment:
I think this can also be done in a simpler way:
```suggestion
    season_groups <- purrr::accumulate(unique_seasons, `c`)
```

Reply:
But I am not sure it gives the exact same result.